libs/src/services/solanaWeb3Manager/rewardsAttester.js

```diff
@@ -0,0 +1,421 @@
+    // coolDown = min(cooldownMsec * backoffExponent ^ retryCount, maxCooldownMsec)
+    this.backoffExponent = 1.5
+    // Maximum time to wait before retrying
+    this.maxCooldownMsec = 10000
+    // Maximum number of retries before moving on
+    this.maxRetries = 5
+
+    this._performSingleAttestation = this._performSingleAttestation.bind(this)
+    this._disbursementToKey = this._disbursementToKey.bind(this)
+  }
+
+  /**
+   * Begin attestation loop.
+   *
+   * @memberof RewardsAttester
+   */
+  async start () {
+    this.logger.info(`Starting attester with:
+      quorum size: ${this.quorumSize}, \
+      parallelization: ${this.parallelization} \
+      AAO endpoint: ${this.aaoEndpoint} \
+      AAO address: ${this.aaoAddress}
+    `)
+    await this._selectDiscoveryNodes()
+
+    while (true) {
+      try {
+        await this._attestInParallel()
+      } catch (e) {
+        this.logger.error(`Got error: ${e}, sleeping`)
+        await this._delay(1000)
+      }
+    }
+  }
+
+  /**
+   * Main method to attest for a bucket of challenges in parallel.
+   *
+   * Algorithm:
+   * - Gets `this.parallelization` undisbursed challenges from the queue, refilling it from DN if necessary.
+   * - Call `_performSingleAttestation` on those in parallel.
+   * - For challenges that failed, either keep retrying or discard them, depending on the error.
+   * - Set offset and startingBlock
+   *
+   * @memberof RewardsAttester
+   */
+  async _attestInParallel () {
+    this.logger.info(`Attesting in parallel with startingBlock: ${this.startingBlock}, offset: ${this.offset}, parallelization: ${this.parallelization}`)
+
+    // Refill queue if necessary, returning early if error
+    const { error } = await this._refillQueueIfNecessary()
+    if (error) {
+      this.logger.error(`Got error trying to refill challenges: [${error}]`)
+      throw new Error(error)
+    }
+
+    // If queue is still empty, sleep and return
+    if (!this.undisbursedQueue.length) {
+      this.logger.info(`No undisbursed challenges. Sleeping...`)
+      await this._delay(1000)
+      return
+    }
+
+    // Get undisbursed rewards
+    let toAttest = this.undisbursedQueue.splice(0, this.parallelization)
+    const highestBlock = Math.max(...toAttest.map(e => e.completedBlocknumber))
+
+    // Attempt to attest in a single sweep
+    const results = await Promise.all(toAttest.map(this._performSingleAttestation))
+
+    // "Process" the results of attestation into noRetry and needsRetry errors,
+    // as well as a flag that indicates whether we should reselect.
+    let { successful, noRetry, needsRetry, shouldReselect } = await this._processResponses(results)
+    let successCount = successful.length
+
+    // Increment offset by the # of errors we're not retrying that have the max block #.
+    //
+    // Note: any successfully completed rewards will eventually be flushed from the
+    // disbursable queue on DN, but ignored rewards will stay stuck in that list, so we
```

Comment:
might be a slightly separate question but will those ignored rewards ever get flushed in the disbursable queue on DN?

Reply:
sorta-kinda: once we move past that block, we won't see them in the results we query because they'd be filtered out by the higher startingBlock. technically they'll always be marked as undisbursed in the db, though